font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,168 @@
 // limitations under the License.
 //
 //! Rasterization powered by DirectWrite
-use dwrote::{FontCollection, FontStretch, FontStyle, FontWeight, GlyphOffset, GlyphRunAnalysis};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::fmt::{self, Display, Formatter};
+use std::os::windows::ffi::OsStringExt;
+
+use dwrote::{
+    FontCollection, FontFace, FontFallback, FontStretch, FontStyle, FontWeight, GlyphOffset,
+    GlyphRunAnalysis, TextAnalysisSource, TextAnalysisSourceMethods, DWRITE_GLYPH_RUN,
+};
+
+use winapi::shared::ntdef::{HRESULT, LOCALE_NAME_MAX_LENGTH};
+use winapi::um::dwrite;
+use winapi::um::winnls::GetUserDefaultLocaleName;
 
 use super::{
     BitmapBuffer, FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight,
 };
 
+/// This struct contains the properties from dwrote::Font needed in the rest of this module.
+///
+/// We use this struct rather than dwrote::Font for performance reasons: the getter methods on
+/// dwrote::Font need to do things like String conversion from C++ widestrings to Rust's
+/// UTF-8 strings.
+pub struct Font {
+    face: FontFace,
+    family_name: String,
+    weight: FontWeight,
+    style: FontStyle,
+    stretch: FontStretch,
+}
+
 pub struct DirectWriteRasterizer {
-    fonts: Vec<dwrote::FontFace>,
+    fonts: HashMap<FontKey, Font>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    collection: FontCollection,
```

Comment:
This is the font fallback list, correct? I think `collection` might not be ideal naming since a collection might be all kinds of things. On Linux this is called `fallback_lists` for example (since every style has a different fallback list).

Or is this a collection of all fonts present on the system?

Reply:
This is the collection of fonts the use. Usually the system collection - and that's what's loaded in `::new()`.

(The dwrote `FontCollection` is a wrapper type around the [DirectWrite interface](https://docs.microsoft.com/en-gb/windows/win32/api/dwrite/nn-dwrite-idwritefontcollection?redirectedfrom=MSDN))

This used to be called `font_collection` but we took the prefix `font_` off most names in the file.

`dwrite_collection` ?

Reply:
Usually when I think collection, I'm directed more torwards `std` than directwrite.

I think `available_fonts` would be a good choice, it seems to correctly represent what is stored.